src/node/audio_buffer_source.rs

```diff
@@ -0,0 +1,673 @@
+//  maybe another macro could also help for Getters and Setters on the node itself
+#[derive(Clone, Debug)]
+struct SharedAttributes {
+    start: Arc<AtomicF64>,
+    stop: Arc<AtomicF64>,
+    offset: Arc<AtomicF64>,
+    duration: Arc<AtomicF64>,
+    loop_: Arc<AtomicBool>,
+    loop_start: Arc<AtomicF64>,
+    loop_end: Arc<AtomicF64>,
+}
+
+impl Default for SharedAttributes {
+    fn default() -> Self {
+        Self {
+            start: Arc::new(AtomicF64::new(0.)),
+            stop: Arc::new(AtomicF64::new(f64::MAX)),
+            offset: Arc::new(AtomicF64::new(0.)),
+            duration: Arc::new(AtomicF64::new(f64::MAX)),
+            loop_: Arc::new(AtomicBool::new(false)),
+            loop_start: Arc::new(AtomicF64::new(0.)),
+            loop_end: Arc::new(AtomicF64::new(0.)),
+        }
+    }
+}
+
+impl SharedAttributes {
+    fn get_start(&self) -> f64 {
+        self.start.load()
+    }
+    fn set_start(&self, value: f64) {
+        self.start.store(value);
+    }
+
+    fn get_stop(&self) -> f64 {
+        self.stop.load()
+    }
+
+    fn set_stop(&self, value: f64) {
+        self.stop.store(value);
+    }
+
+    fn get_offset(&self) -> f64 {
+        self.offset.load()
+    }
+
+    fn set_offset(&self, value: f64) {
+        self.offset.store(value);
+    }
+
+    fn get_duration(&self) -> f64 {
+        self.duration.load()
+    }
+
+    fn set_duration(&self, value: f64) {
+        self.duration.store(value);
+    }
+
+    fn get_loop(&self) -> bool {
+        self.loop_.load(Ordering::SeqCst)
+    }
+
+    fn set_loop(&self, value: bool) {
+        self.loop_.store(value, Ordering::SeqCst);
+    }
+
+    fn get_loop_start(&self) -> f64 {
+        self.loop_start.load()
+    }
+
+    fn set_loop_start(&self, value: f64) {
+        self.loop_start.store(value);
+    }
+
+    fn get_loop_end(&self) -> f64 {
+        self.loop_end.load()
+    }
+
+    fn set_loop_end(&self, value: f64) {
+        self.loop_end.store(value);
+    }
+}
+
+pub struct AudioBufferSourceOptions {
+    pub buffer: Option<AudioBuffer>,
+    pub detune: f32,
+    pub loop_: bool,
+    pub loop_start: f64,
+    pub loop_end: f64,
+    pub playback_rate: f32,
+    pub channel_config: ChannelConfigOptions,
+}
+
+impl Default for AudioBufferSourceOptions {
+    fn default() -> Self {
+        Self {
+            buffer: None,
+            detune: 0.,
+            loop_: false,
+            loop_start: 0.,
+            loop_end: 0.,
+            playback_rate: 1.,
+            channel_config: Default::default(),
+        }
+    }
+}
+
+// channels, duration
+// @todo - define if we could pass an `AudioBuffer` directly
+struct BufferChannelsMessage(Vec<Arc<Vec<f32>>>, f64);
+
+pub struct AudioBufferSourceNode {
+    registration: AudioContextRegistration,
+    attributes: SharedAttributes,
+    channel_config: ChannelConfig,
+    sender: Sender<BufferChannelsMessage>,
+    detune: AudioParam,        // has constraints, no a-rate
```

Comment:
These two params are not actually used in rendering, am I right (also not in the current implementation)

Reply:
yes they are see `audio_buffer_source` line 427

Reply:
the `computed_playback_rate` is calculated here  https://github.com/b-ma/web-audio-api-rs/blob/a60ce47ea5faa8fcffcf4b01f4d74e72ded09517/src/node/audio_buffer_source.rs#L431 and used to update the renderer internal state https://github.com/b-ma/web-audio-api-rs/blob/a60ce47ea5faa8fcffcf4b01f4d74e72ded09517/src/node/audio_buffer_source.rs#L576 and https://github.com/b-ma/web-audio-api-rs/blob/a60ce47ea5faa8fcffcf4b01f4d74e72ded09517/src/node/audio_buffer_source.rs#L577